iceaxe/__tests__/conftest.py

```diff
@@ -1,84 +1,300 @@
+                            f"Failed to stop container {container.name}: {e}"
+                        )
+
+    # Double-check if the port is free now
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.bind(("0.0.0.0", port))
+            return True
+    except OSError:
+        return False
+
+
+@pytest.fixture(scope="session")
+def docker_postgres():
+    """
+    Fixture that creates a PostgreSQL container using the Python Docker API.
+    This allows running individual tests without needing Docker Compose.
+    """
+    # Initialize Docker client
+    client = docker.from_env()
+
+    # Generate a unique container name to avoid conflicts
+    container_name = f"postgres-test-{uuid.uuid4().hex[:8]}"
+
+    # First try with the fixed port to match existing test expectations
+    preferred_port = 5438
+
+    # Try to free up the preferred port
+    port_freed = stop_containers_using_port(client, preferred_port)
+
+    # If we couldn't free the port, use a random one
+    if port_freed:
+        port = preferred_port
+        logger.info(f"Using preferred port {port}")
+    else:
+        port = get_free_port()
+        logger.info(f"Using alternative port {port}")
+
+    # PostgreSQL connection details
+    pg_user = "iceaxe"
+    pg_password = "mysecretpassword"
+    pg_db = "iceaxe_test_db"
+
+    # Start PostgreSQL container
+    try:
+        container = client.containers.run(
+            "postgres:16",
+            name=container_name,
+            detach=True,
+            environment={
+                "POSTGRES_USER": pg_user,
+                "POSTGRES_PASSWORD": pg_password,
+                "POSTGRES_DB": pg_db,
+            },
+            ports={"5432/tcp": port},
+            remove=True,  # Auto-remove container when stopped
+        )
+    except APIError as e:
+        # If there's still an issue, try with a random port as a last resort
+        if "port is already allocated" in str(e):
+            logger.warning(
+                f"Port {port} is still in use. Trying with a completely random port."
+            )
+            port = get_free_port()
+            try:
+                container = client.containers.run(
+                    "postgres:16",
+                    name=container_name,
+                    detach=True,
+                    environment={
+                        "POSTGRES_USER": pg_user,
+                        "POSTGRES_PASSWORD": pg_password,
+                        "POSTGRES_DB": pg_db,
+                    },
+                    ports={"5432/tcp": port},
+                    remove=True,
+                )
+            except Exception as inner_e:
+                raise RuntimeError(
+                    f"Failed to start PostgreSQL container with random port: {inner_e}"
+                ) from e
+        else:
+            raise RuntimeError(f"Failed to start PostgreSQL container: {e}")
+
+    # Wait for PostgreSQL to be ready
```

Comment:
This function's getting a bit long for this runtime behavior. Feels like it could benefit from refactoring the docker launcher / status checking code into a class or different functions.